editor/src/components/editor/store/collaborative-editing.ts

```diff
@@ -0,0 +1,363 @@
+          }
+          synchroniseParseSuccessToCollabFile(parsedPart, collabFile)
+        })
+      }
+      break
+    case 'ENSURE_DIRECTORY_EXISTS':
+      // Not necessary to handle right now.
+      break
+    default:
+      assertNever(change)
+  }
+}
+
+export function updateCollaborativeProjectContents(
+  session: CollaborativeEditingSupportSession,
+  projectChanges: ProjectChanges,
+  filesModifiedByAnotherUser: Array<string>,
+): void {
+  const projectContentsMap = session.projectContents
+  for (const change of projectChanges.fileChanges.collabProjectChanges) {
+    if (!filesModifiedByAnotherUser.includes(change.fullPath)) {
+      applyFileChangeToMap(change, projectContentsMap, session.mergeDoc)
+    }
+  }
+}
+
+export function addHookForProjectChanges(
+  session: CollaborativeEditingSupportSession,
+  dispatch: EditorDispatch,
+): void {
+  session.projectContents.observeDeep((changeEvents) => {
+    // TODO Check that this is the array change before doing anything
+    for (const changeEvent of changeEvents) {
+      switch (changeEvent.path.length) {
+        // This case indicates a change at the base of the entire structure, which
+        // appears to arise at least on first connection to sync up the entire value.
+        case 0: {
+          if (changeEvent instanceof Y.YMapEvent) {
+            updateEntireProjectContents(changeEvent as Y.YMapEvent<any>, dispatch)
+          } else {
+            throw new Error(`Could not treat change event as Y.YMapEvent.`)
+          }
+          break
+        }
+        // Originally thought to be a case that would arise on a new addition of
+        // a file, left here to capture this specific case in case it does show up.
+        case 1: {
+          console.error(`Unhandled path length of 1: ${changeEvent.path}`)
+          break
+        }
+        // When a change happens to the `topLevelElements` in a particular file,
+        // this case should show up as the path will consist of the filename and
+        // the string `topLevelElements`.
+        case 2: {
+          const filePath = changeEvent.path[0] as string
+          if (changeEvent.path[1] !== 'topLevelElements') {
+            throw new Error(`Unexpected second part of change path: ${changeEvent.path[1]}`)
+          }
+          updateTopLevelElementsOfFile(session, filePath, changeEvent, dispatch)
+          break
+        }
+        default:
+          throw new Error(`Unexpected change path: ${JSON.stringify(changeEvent.path)}`)
+      }
+    }
+  })
+}
+
+export interface ArrayChanges {
+  updatesAt: Array<number>
+  deleteFrom: number | null
+}
+
+function updateEntireProjectContents(
+  changeEvent: Y.YMapEvent<any>,
+  dispatch: EditorDispatch,
+): void {
+  let actions: Array<EditorAction> = []
+  const entriesIterator = changeEvent.keys.entries()
+  let entriesIteratorResult = entriesIterator.next()
+  while (!entriesIteratorResult.done) {
```

Comment:
I believe we could use a `for (const [filename, changeEntry] of entriesIterator) {` here

Reply:
That's much cleaner, changed in 21cee15c7.